srcs/app/http/HttpRequest.cpp

```diff
@@ -0,0 +1,195 @@
+	}
+	return false;
+}
+
+bool	HttpRequest::ParseUri_(const std::string &raw_request) {
+	const std::size_t	uri_start = offset_;
+
+	offset_ = raw_request.find(' ', uri_start);
+	if (offset_ == std::string::npos)
+		return false;
+	uri_ = raw_request.substr(uri_start, offset_ - uri_start);
+	++offset_;
+	return IsValidUri_(uri_);
+}
+
+bool	HttpRequest::IsValidUri_(const std::string &uri) const {
+	// TODO(gbudau) Look into how to validate URI
+	(void)uri;
+	return true;
+}
+
+bool	HttpRequest::ParseHttpVersion_(const std::string &raw_request) {
+	const std::size_t	http_version_start = offset_;
+
+	offset_ = raw_request.find(kCRLF_, http_version_start);
+	if (offset_ == std::string::npos)
+		return false;
+	http_version_ = raw_request.substr(http_version_start,
+										offset_ - http_version_start);
+	offset_ += 2;
+	return IsValidHttpVersion_(http_version_);
+}
+
+bool	HttpRequest::IsValidHttpVersion_(const std::string &http_version) const {
+	return http_version == "HTTP/1.1";
+}
+
+bool	HttpRequest::ParseHeaders_(const std::string &raw_request) {
+	while (offset_ < raw_request.length()) {
+		const std::size_t header_end = raw_request.find(kCRLF_, offset_);
+		if (header_end == std::string::npos)
+			return false;
+		if (header_end - offset_ == 0)
+			break;
+		HeaderName name = ParseHeaderName_(raw_request);
+		if (name.empty())
+			return false;
+		HeaderValue value = ParseHeaderValue_(raw_request);
+		if (value.empty())
+			return false;
+		AddHeader_(name, value);
+		offset_ += 2;
+	}
+	offset_ += 2;
+	return HasHeader("Host");
+}
+
+HttpRequest::HeaderName
+HttpRequest::ParseHeaderName_(const std::string &raw_request) {
+	const std::size_t name_start = offset_;
+
+	offset_ = raw_request.find(':', name_start);
+	if (offset_ == std::string::npos)
+		return "";
+	HeaderName	name = raw_request.substr(name_start, offset_ - name_start);
+	if (!IsValidHeaderName_(name))
+		return "";
+	name = ToLowerString(name);
+	++offset_;
+	return name;
+}
+
+HttpRequest::HeaderValue
+HttpRequest::ParseHeaderValue_(const std::string &raw_request) {
+	const std::size_t	value_start = offset_;
+
+	offset_ = raw_request.find(kCRLF_, offset_);
+	if (offset_ == std::string::npos)
+		return "";
+	HeaderValue value = raw_request.substr(value_start, offset_ - value_start);
+	value = TrimString(value, kWhitespace_);
+	if (!IsValidHeaderValue_(value))
+		return "";
+	return value;
+}
+
+void	HttpRequest::AddHeader_(
+		const HeaderName &name, const HeaderValue &value) {
+	headers_.insert(std::make_pair(name, value));
+}
+
+// https://datatracker.ietf.org/doc/html/rfc7230#section-3.2.6
+bool	HttpRequest::IsValidHeaderName_(const std::string &name) const {
+	const std::string valid_chars = "!#$%&'*+-.^_`|~"
+								"0123456789"
+								"abcdefghijklmnopqrstuvwxyz"
+								"ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+
+	return !name.empty() &&
+		name.find_first_not_of(valid_chars) == std::string::npos;
+}
+
+bool	HttpRequest::IsValidHeaderValue_(const std::string &value) const {
```

Comment:
We may have a more intensive header error checking? for example, something like `"Accept-Language: paco\r\n"` would pass your tests. Maybe we should investigate if this returns a `BAD REQUEST` error or if it just ignores it.

Reply:
Good idea. For now, I added this to a new task in #14 until we know more about what headers we want to implement.